community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/pipes/ExpandIntoPipe.scala

```diff
@@ -148,19 +151,29 @@ case class ExpandIntoPipe(source: Pipe,
   def withEstimatedCardinality(estimated: Double) = copy()(Some(estimated))
 
   private final class RelationshipsCache(capacity: Int) {
-    private var _size = 0
-    private val table = new mutable.OpenHashMap[(Node, Node), Seq[Relationship]]()
-    def put(node1: Node, node2: Node, rels: Seq[Relationship]) = {
-      if (_size < capacity) {
-        table.put(key(node1, node2), rels)
-        _size += 1
+
+    val table = new mutable.OpenHashMap[(Long, Long), Seq[Relationship]]()
+
+    def put(start: Node, end: Node, rels: Seq[Relationship]) = {
+      if (table.size < capacity) {
+        table.put(key(start, end), rels)
       }
     }
 
-    def get(node1: Node, node2: Node) = table.get(key(node1, node2))
+    def recordNoRels(start: Node, end: Node) = {
+      put(start, end, RelationshipsCache.NoRels)
+      put(end, start, RelationshipsCache.NoRels)
+    }
+
+    def get(start: Node, end: Node) = table.get(key(start, end))
 
     @inline
-    private def key(node1: Node, node2: Node) =
-      if (node1.getId < node2.getId) (node1, node2) else (node2, node1)
+    private def key(start: Node, end: Node) = (start.getId, end.getId)
```

Comment:
Changed code a bit. For `Direction.BOTH` ordering in tuple is not important so we sort elements. For `Direction.OUTGOING` and `Direction.INCOMING` ordering is important - cache should serve different sets of relationships for `(Node1, Node2)` and `(Node2, Node1)`.